src/test/spec/unified_runner/test_event.rs

```diff
@@ -1,38 +1,191 @@
-use crate::{bson::Document, test::CommandEvent};
+use crate::{
+    bson::Document,
+    event::cmap::{ConnectionCheckoutFailedReason, ConnectionClosedReason},
+    test::{CmapEvent, CommandEvent, Event},
+};
 use serde::Deserialize;
 
 #[derive(Debug, Deserialize, PartialEq)]
-pub enum TestEvent {
-    #[serde(rename = "commandStartedEvent")]
+#[serde(untagged, deny_unknown_fields, rename_all = "camelCase")]
+pub enum ExpectedEvent {
+    Cmap(ExpectedCmapEvent),
+    Command(ExpectedCommandEvent),
+    Sdam,
+}
+
+impl From<Event> for ExpectedEvent {
+    fn from(event: Event) -> Self {
+        match event {
+            Event::Command(sub) => ExpectedEvent::Command(sub.into()),
+            Event::Cmap(sub) => ExpectedEvent::Cmap(sub.into()),
+            Event::Sdam(_) => ExpectedEvent::Sdam,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(deny_unknown_fields)]
+pub enum ExpectedCommandEvent {
+    #[serde(rename = "commandStartedEvent", rename_all = "camelCase")]
     Started {
         command_name: Option<String>,
         database_name: Option<String>,
         command: Option<Document>,
+        has_service_id: Option<bool>,
     },
-    #[serde(rename = "commandSucceededEvent")]
+    #[serde(rename = "commandSucceededEvent", rename_all = "camelCase")]
     Succeeded {
         command_name: Option<String>,
         reply: Option<Document>,
+        has_service_id: Option<bool>,
+    },
+    #[serde(rename = "commandFailedEvent", rename_all = "camelCase")]
+    Failed {
+        command_name: Option<String>,
+        has_service_id: Option<bool>,
     },
-    #[serde(rename = "commandFailedEvent")]
-    Failed { command_name: Option<String> },
 }
 
-impl From<CommandEvent> for TestEvent {
+impl From<CommandEvent> for ExpectedCommandEvent {
```

Comment:
Idea for future cleanup: it might be worthwhile to remove the `From<Event>` machinery and directly match `ExpectedEvent` against `Event`.

Reply:
Yeah I agree. As it is right now it's a little confusing that the `actual_event` in `events_match` is an `ExpectedEvent` type.